### PR TITLE
T11983

### PR DESCRIPTION
**Pull request by manuq**
https://phabricator.endlessm.com/T20365

### CODE REVIEW

File: devscripts/bashcompletion.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+here = os.path.abspath(os.path.dirname(__file__))
+sys.path.insert(0, os.path.join(here, '..', 'flapjack'))
+
+from flapjack import commands
+
+BASH_COMPLETION_TEMPLATE = os.path.join(here, "bash-completion.in")
+BASH_COMPLETION_FILE = os.path.join(here, '..', 'dist', "flapjack.bash-completion")
+
+def get_command_vars():
+    all_commands = commands._command_registry.keys()
```
Comment (ptomato) [inline devscripts/bashcompletion.py]: You could move this inside `commands.py` and thereby avoid accessing private variables from here. (In fact I looked at how flatpak does it, and they have a `flatpak complete` command which doesn't show up in the help, but you can see it for example with `flatpak complete 'flatpak build' 13`.

File: setup.py [hunk 2]
```diff
@@ -81,6 +81,24 @@ def run_tests(self):
             sys.exit(0)  # no tests is okay, since we don't have any tests yet
         sys.exit(exitcode)
 
+class BuildCompletionCommand(Command):
+    """Support setup.py build_completion to build the bash completion script."""
+
+    description = 'Build the bash completion script'
+    user_options = []
+
+    def initialize_options(self):
```
Comment (ptomato) [inline setup.py]: These are not needed if they don't do anything, I think.

File: setup.py [hunk 3]
```diff
@@ -81,6 +81,24 @@ def run_tests(self):
             sys.exit(0)  # no tests is okay, since we don't have any tests yet
         sys.exit(exitcode)
 
+class BuildCompletionCommand(Command):
```
Comment (ptomato) [inline setup.py]: It might be nice to have this done as part of `setup.py build`.

**Comment by ptomato** [comment]
Sorry for the delay. Looks good. It might be handy to have a `setup.py install_completions` command which would stick the bash script in the right place (and likely have to be run with sudo.)

**Review by ptomato — CHANGES_REQUESTED** [review]

File: devscripts/bashcompletion.py (hunk 1 above)
Comment (manuq) [reply devscripts/bashcompletion.py]: Aha! I looked inside the flatpak code for inspiration and found the hidden `flatpak complete` command, but I couldn't figure the arguments to run it.
I think I'll switch to create a `flapjack complete` hidden command, and move the logic to Python. The bash autocomplete script would then be static and simpler, no need to build it, just to install it.
The only inconvenience is Python infamous startup time, as flatpak will be called each time TAB is pressed to complete, I will keep an eye on that.

File: setup.py [hunk 4]
```diff
@@ -27,6 +31,17 @@
     exec(f.read(), about)
 
 
+def _get_data_files():
+    pkg_config_command = 'pkg-config --variable=completionsdir bash-completion'
+    bash_completions_dir = subprocess.getoutput(pkg_config_command)
+    if bash_completions_dir:
+        return [(bash_completions_dir, ['build/flapjack.bash-completion'])]
```
Comment (ptomato) [inline setup.py]: I'm not sure this does what it appears to do. Googling suggests that it installs the file inside the EGG. In any case when I install with `pip3 install --user .` it doesn't put the completion file in `/usr/share/bash-completion/completions`.

In addition, the other files in my completions directory don't seem to have the `.bash-completion` extension, does that vary from system to system?

**Review by ptomato — CHANGES_REQUESTED** [review]
You may need to add a rule to `debian/rules` to get the completions file in the right place in the Debian package.

File: setup.py [hunk 5]
```diff
@@ -119,5 +144,7 @@ def run_tests(self):
     cmdclass={
         'publish': PublishCommand,
         'test': PyTestCommand,
+        'build_py': BuildCompletionCommand,
     },
+    data_files=data_files,
```
Comment (ptomato) [inline setup.py]: Small nitpick: I would just put `data_files=_get_data_files()` here.

File: setup.py (hunk 4 above)
Comment (ptomato) [reply setup.py]: Did you see this comment from last time?

**Review by ptomato — CHANGES_REQUESTED** [review]
I don't have Debian packaging tools on my OSTree system so I can't test that part of it. Did it include the completion file for you?

File: devscripts/bashcompletion.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+here = os.path.abspath(os.path.dirname(__file__))
+sys.path.insert(0, os.path.join(here, '..'))
+
+
+from flapjack import commands  # noqa
+
+
+BASH_COMPLETION_TEMPLATE = os.path.join(here, "bash-completion.in")
+BASH_COMPLETION_FILE = os.path.join(here, '..', 'build',
+                                    "flapjack.bash-completion")
```
Comment (ptomato) [inline devscripts/bashcompletion.py]: Nitpick: be consistent with single-quote strings.

File: setup.py [hunk 7]
```diff
@@ -82,6 +83,51 @@ def run_tests(self):
...
+
+    def finalize_options(self):
+        pass
+
+    def run(self):
+        pkg_config_command = ('pkg-config --variable=completionsdir' +
+                              'bash-completion')
+        bash_completions_dir = subprocess.getoutput(pkg_config_command)
+        if not bash_completions_dir:
+            print("Couldn't install completions: No completions dir found.")
```
Comment (ptomato) [inline setup.py]: Ditto.

File: setup.py [hunk 8]
```diff
@@ -82,6 +83,51 @@ def run_tests(self):
...
+            sys.exit(1)
+
+        completions_file = "build/flapjack.bash-completion"
```
Comment (ptomato) [inline setup.py]: Ditto.

**Review by ptomato — CHANGES_REQUESTED** [review]
Looks fine. Up to you if you want to fix the quotes and merge it as-is, opening a follow-up issue for the debian/rules changes that we discussed, or leave it until you get a chance to do that.

**Comment by manuq** [comment]
@ptomato please note that I made the last commit by only looking at the Debian documentation, but is UNTESTED. I will ask someone with Debian tools to test it. This is not my case.
 - https://www.debian.org/doc/manuals/maint-guide/dreq.en.html#customrules 
 - https://manpages.debian.org/jessie/bash-completion/dh_bash-completion.1.en.html

**Comment by ptomato** [comment]
Maybe the best way to test it is to get OBS to build it...

**Review by ptomato — APPROVED** [review]

**Comment by ptomato** [comment]
I merged it manually.

**Pull request closed without merging**